File: src/main/java/net/tassia/ini/IniParser.java
package net.tassia.ini;

import net.tassia.util.Parser;

public class IniParser extends Parser {

	public IniParser(String source) {
		super(source);
	}



	public IniDocument readDocument() {
		IniDocument doc = new IniDocument();
		IniSection section = doc.getRoot();

		// Main loop
		while (true) {

			skipWhitespace();
			if (hasEndReached()) break;

			// Comment?
			if (isNext('#') || isNext(';')) {
				readUntil('\n');
				continue;
			}

			// Section
			if (isNext('[')) {
				String sectionName = readUntil(']').trim();
				if (!isNext(']')) throw new IllegalArgumentException("Invalid INI file format.");
				section = doc.getSection(sectionName);
				continue;
			}

			// Key
			String key = readWhile((c) -> c != '=' && c != '\n').trim();
			String value = "";
			if (isNext('=')) {
				value = readWhile((c) -> c != '\n').trim();
			}
			section.put(key, value);

		}

		return doc;
	}

}

File: kommons-ini/src/commonMain/kotlin/net.tassia.ini/IniParser.kt
package net.tassia.ini

import net.tassia.parser.Parser

fun Parser.readIniDocument(): IniDocument {

	val doc = IniDocument()
	var section = doc.root

	// Main loop
	while (true) {

		skipWhitespace()
		if (hasEndReached) break

		// Comment?
		if (isNext('#') || isNext(';')) {
			readUntil('\n')
			continue
		}

		// Section
		if (isNext('[')) {
			val sectionName = readUntil(']').trim()
			if (!isNext(']')) throw IllegalArgumentException("Invalid INI file format.")
			section = doc.section(sectionName)
			continue
		}

		// Key
		val key = readWhile { it != '=' && it != '\n' }.trim()
		var value = ""
		if (isNext('=')) {
			value = readWhile { it != '\n' }.trim()
		}
		section[key] = value

	}

	return doc

}
